### ISSUE TITLE
Restore the slime-presentation-streams contrib

### ISSUE THREAD

**Issue by luismbo** [slime/slime#213]
It was disabled due to issue #212. https://sourceforge.net/p/sbcl/mailman/message/33102884/ outlines how to reimplement it properly.

**Comment by luismbo** [comment]
@phmarek fixed some newer breakage here: https://github.com/phmarek/slime/commit/e3c3540786a000d7afee1cd420b99c3ba3c5f7f7

**Comment by luismbo** [comment]
@snuglas how likely do you think it would be for SLIME to have some way to hook into pretty printing? Would it help to send an e-mail to sbcl-devel (or similar) describing slime-presentation-streams's needs?

**Comment by scymtym** [comment]
Another data point for that discussion would be https://github.com/McCLIM/McCLIM/tree/master/Experimental/pprint.

I think it also tried to turn pieces of pretty print output into presentations.

CC @dkochmanski

**Comment by snuglas** [comment]
probably best to ask on sbcl-devel, even better if you can also propose
some mechanism by which to achieve the outcome.

>

**Comment by luismbo** [comment]
@phmarek, @scymtym: are any of you inclined to propose an API and/or patch?

**Comment by scymtym** [comment]
> @scymtym: are any of you inclined to propose an API and/or patch?

I don't have enough time at the moment, sorry.

**Comment by phmarek** [comment]
Well, why not take _stassats_ idea from the ML? Or does that solve a different problem?

**Comment by luismbo** [comment]
@stassats has [your idea](https://sourceforge.net/p/sbcl/mailman/message/33102884/) matured in the past 4 years? :-)

**Comment by stassats** [comment]
Yes, it's prime for somebody (read _not me_) to implement it and integrate it into Slime.

**Comment by luismbo** [comment]
@stassats I wonder if it'd compose better to have make-pretty-stream be a generic function that dispatches on the target stream? That way we could add methods for SLIME streams and leave room for other prettiness hacks without setting a global variable.

**Comment by stassats** [comment]
It's not too early for CLOS (i.e. when Slime is loaded). The target stream would have to be subclassed and methods are easily dynamically scoped.

**Comment by luismbo** [comment]
BTW, I recently noticed that `swank-presentation-streams` has two mechanisms for hooking into object printing, one of which is only used when in `*use-dedicated-output-stream*` mode that is much less intrusive and doesn't require hooking into `make-pretty-stream`. I think this explains how @phmarek is using this contrib when I had the impression that the contrib didn't work at all.

**Comment by luismbo** [comment]
@stassats sorry, I don't understand what you mean. I was thinking that the constructor for `pretty-stream` could be named `%make-pretty-stream` or something and a generic function `make-pretty-stream` would dispatch on the target stream and either call `%make-pretty-stream` (a default method provided by SBCL) or our `make-pretty-slime-stream` (via our own method). What did you have in mind?
